Enable map key sorting in `utils.FastMarshal`

timothyb89 opened the pull request:
This enables map key sorting during JSON serialization to ensure `CompareAndSwap` opertions work consistently if the resource in question has embedded map values (e.g. labels).

Note that this is hypothetically a breaking change: it makes comparisons consistent, so if it saves a variant that consistently fails a `CompareAndSwap()`, it will... fail consistently rather than sometimes work. In practice, today only `CertificateAuthority` resources use `CompareAndSwap()` and under normal circumstances they never have labels attached.

In the unlikely event that some resource is affected, a load/store should save it with sorted key ordering and fix the problem.

timothyb89 committed:
Enable map key sorting in `utils.FastMarshal`

This enables map key sorting during JSON serialization to ensure
`CompareAndSwap` opertions work consistently if the resource in
question has embedded map values (e.g. labels).

Note that this is hypothetically a breaking change: it makes
comparisons consistent, so if it saves a variant that consistently
fails a `CompareAndSwap()`, it will... fail consistently rather
than sometimes work. In practice, today only `CertificateAuthority`
resources use `CompareAndSwap()` and under normal circumstances they
never have labels attached.

In the unlikely event that some resource is affected, a load/store
should save it with sorted key ordering and fix the problem.

nklaassen commented:
I _think_ that `CompareAndSwapCertAuthority` is the only use of `CompareAndSwap`, and I **really** hope that nothing else is comparing serialized strings in the backend.

I am pretty sure that we have never set any labels on `CertAuthorities`, but to play it safe you could add a check to `CompareAndSwapCertAuthority` that no labels are set. If there are, since I don't expect to see this, logging a helpful warning is probably enough. In theory you could get creative with backend locks and write the sorted version to the backend but that's probably overkill.

espadolini commented:
I _might_ be adding a label in the metadata of `CertAuthority` soon; perhaps the warning could be added whenever `len(ca.Metadata.Labels) >= 2`? There should be no ordering problem when it's just one key, right?

timothyb89 committed:
Log a warning if CAs fail a compare-and-swap with multiple labels

This attempts to mitigate compare-and-swap failures if CAs have
multiple labels by logging a warning instructing users to re-save
their CertificateAuthority resource.

timothyb89 commented:
That's reasonable, it now logs a warning if comparisons fail and multiple labels are present.

espadolini commented on lib/utils/jsontools.go:
```diff
@@ -70,10 +70,19 @@ func FastUnmarshal(data []byte, v interface{}) error {
+// sorting to ensure CompareAndSwap checks consistently succeed.
+var SafeConfig = jsoniter.Config{
+	EscapeHTML:                    false,
+	MarshalFloatWith6Digits:       true, // will lose precession
```
[Precession is a change in the orientation of the rotational axis of a rotating body.](https://en.wikipedia.org/wiki/Precession)

I realize that this is taken right from the definition of `jsoniter.ConfigFastest` but maybe we don't need to also mirror the misspelled comments 😅

timothyb89 replied:
oh geez I'm embarrassed that I missed the typo, will fix!

zmb3 commented on lib/services/local/trust.go:
```diff
@@ -121,6 +122,9 @@ func (s *CA) CompareAndSwapCertAuthority(new, existing types.CertAuthority) erro
 	if err != nil {
 		if trace.IsCompareFailed(err) {
+			if len(existing.GetMetadata().Labels) >= 2 {
+				log.Warn("compare-and-swap comparison failed on certificate authority with multiple labels; it may need to be re-saved to sort map keys")
```
Any warnings or errors we log should be actionable and easy to understand by an end user, not just developers.

If a user sees this warning message, what do we expect them to do about it?

- if we don't expect them to do anything, then it's probably not a warning
- if we do expect them to do something, let's make the message more clear as to _what_ that is

timothyb89 replied:
Fair point, I've changed it to now output a message like the following:

> comparison failed on certificate authority with multiple labels; if this occurs consistently, try re-saving the resource: tctl get cert_authority/user/foo.example.com --with-secrets > ca.yaml && tctl create -f ca.yaml && rm ca.yaml

timothyb89 committed:
Fix copied comment typo and make the logged warning actionable

fspmarshall reviewed (approved):
LGTM.  Determinism is always good.

For those who are curious, there is a "correct" way to compare and swap under non-deterministic serialization.  CAs just didn't do it for some reason:

1. Load existing value.
2. Compare expected value to *deserialized* version of existing value.
3. Call `Backend.CompareAndSwap` with the value loaded in step 1.

Simply having determinism is better/faster though.

timothyb89 committed:
Fix log.Warn() call with formatting directives

timothyb89 committed:
Merge remote-tracking branch 'origin/master' into timothyb89/cas-sort-map-keys

timothyb89 committed:
Merge branch 'master' into timothyb89/cas-sort-map-keys

timothyb89 committed:
Merge branch 'master' into timothyb89/cas-sort-map-keys
